pkg/restore/prioritize_group_version.go

```diff
@@ -0,0 +1,337 @@
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/version"
+
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/archive"
+	"github.com/vmware-tanzu/velero/pkg/client"
+)
+
+// ChosenGroupVersion is the API Group version that was selected to restore
+// from potentially multiple backed up version enabled by the feature flag
+// APIGroupVersionsFeatureFlag
+type ChosenGroupVersion struct {
+	Group   string
+	Version string
+	Dir     string
+}
+
+// chooseAPIVersionsToRestore will choose a version to restore based on a user-
+// provided config map prioritization or our version prioritization.
+func (ctx *restoreContext) chooseAPIVersionsToRestore() error {
+	sourceGVs, targetGVs, userGVs, err := ctx.gatherSourceTargetUserGroupVersions()
+	if err != nil {
+		return err
+	}
+
+OUTER:
+	for rg, sg := range sourceGVs {
+		// Default to the source preferred version if no other common version
+		// can be found.
+		cgv := ChosenGroupVersion{
+			Group:   sg.Name,
+			Version: sg.PreferredVersion.Version,
+			Dir:     sg.PreferredVersion.Version + velerov1api.PreferredVersionDir,
+		}
+
+		tg := findAPIGroup(targetGVs, sg.Name)
+		if len(tg.Versions) == 0 {
+			ctx.SetChosenGVToRestore(cgv, rg, "", "")
+			continue
+		}
+
+		// Priority 0: User Priority Version
+		if userGVs != nil {
+			uv := findSupportedUserVersion(userGVs[rg].Versions, tg.Versions, sg.Versions)
+			if uv != "" {
+				ctx.SetChosenGVToRestore(cgv, rg, sg.PreferredVersion.Version, uv)
+				ctx.log.Debugf("APIGroupVersionsFeatureFlag Priority 0: User defined API group version %s chosen for %s", uv, rg)
+				continue
+			}
+
+			ctx.log.Infof("Cannot find user defined version in both the cluster and backup cluster. Ignoring version %s for %s", uv, rg)
+		}
+
+		// Priority 1: Target Cluster Preferred Version
+		if versionsContain(sg.Versions, tg.PreferredVersion.Version) {
+			ctx.SetChosenGVToRestore(
+				cgv,
+				rg,
+				sg.PreferredVersion.Version,
+				tg.PreferredVersion.Version,
+			)
+			ctx.log.Debugf(
+				"APIGroupVersionsFeatureFlag Priority 1: Cluster preferred API group version %s found in backup for %s",
+				tg.PreferredVersion.Version,
+				rg,
+			)
+			continue
+		}
+
+		// Priority 2: Source Cluster Preferred Version
+		if versionsContain(tg.Versions, sg.PreferredVersion.Version) {
+			ctx.SetChosenGVToRestore(
+				cgv,
+				rg,
+				sg.PreferredVersion.Version,
+				sg.PreferredVersion.Version,
+			)
+			ctx.log.Debugf(
+				"APIGroupVersionsFeatureFlag Priority 2: Cluster preferred API group version not found in backup. Using backup preferred version %s for %s",
+				sg.PreferredVersion.Version,
+				rg,
+			)
+			continue
+		}
+
+		// Priority 3: The Common Supported Version with the Highest Kubernetes Version Priority
+		for _, tv := range tg.Versions[1:] {
```

Comment:
Short answer: The first elements are the preferred versions. Preferred versions were checked as potential versions to restore in priorities 1 and 2.

Long answer: I am using Kubernetes priority versions ordering. After k8s orders the versions, the first one (the latest, most stable version) will be the preferred. Because I order all the versions as soon as I get them from source, target, and user, the first first will always be the preferred version. And because preferred versions were already considered and rejected by the time we get to the Priority 3 case, I skipped checking them.